File: practice/data_structures/linked_lists/merge_two_sorted_linked_lists/merge_two_sorted_linked_lists.cpp
/*
  Merge two sorted lists A and B as one linked list
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/

Node * MergeLists(Node * i_headA, Node * i_headB)
{
    if (!i_headA)
    {
        return i_headB;
    }
    if (!i_headB)
    {
        return i_headA;
    }
    
    Node * head;
    if (i_headA->data < i_headB->data)
    {
        head = i_headA;
        i_headA = i_headA->next;
    }
    else
    {
        head = i_headB;
        i_headB = i_headB->next;
    }
    
    Node * cp = head;
    while (i_headA && i_headB)
    {
        if (i_headA->data < i_headB->data)
        {
            head->next = i_headA;
            i_headA = i_headA->next;
        }
        else
        {
            head->next = i_headB;
            i_headB = i_headB->next;
        }
        head = head->next;
    }
    
    if (!i_headA && i_headB)
    {
        head->next = i_headB;
    }
    else if (i_headA && !i_headB)
    {
        head->next = i_headA;
    }
    
    return cp;
}

File: practice/data_structures/linked_lists/merge_two_sorted_linked_lists/merge_two_sorted_linked_lists.py
"""
 Merge two linked lists
 head could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None):
       self.data = data
       self.next = next_node

 return back the head of the linked list in the below method.
"""

def MergeLists(i_headA, i_headB):
    if not i_headA:
        return i_headB
    if not i_headB:
        return i_headA
    
    head = Node()
    if i_headA.data < i_headB.data:
        head = i_headA
        i_headA = i_headA.next
    else:
        head = i_headB
        i_headB = i_headB.next
    
    cp = head
    while i_headA and i_headB:
        if i_headA.data < i_headB.data:
            head.next = i_headA
            i_headA = i_headA.next
        else:
            head.next = i_headB
            i_headB = i_headB.next
        head = head.next
    
    if not i_headA and i_headB:
        head.next = i_headB
    elif i_headA and not i_headB:
        head.next = i_headA
    return cp

